core/trino-main/src/main/java/io/trino/metadata/ResolvedFunction.java

```diff
@@ -73,6 +80,7 @@ public ResolvedFunction(
         this.functionNullability = requireNonNull(functionNullability, "nullability is null");
         this.typeDependencies = ImmutableMap.copyOf(requireNonNull(typeDependencies, "typeDependencies is null"));
         this.functionDependencies = ImmutableSet.copyOf(requireNonNull(functionDependencies, "functionDependencies is null"));
+        this.qualifiedName = Suppliers.memoize(this::resolveQualifiedName);
```

Comment:
This commit, besides introducing shared compressor/decompressor, also adds memoization here.
Should this be a separate commit?

BTW is it common that `toQualifiedName` is invoked multiple times on a `ResolvedFunction` instance?
(In case it's always or almost always involved, we could precompute that in constructor, and without need for memoization.)